src/vector/rageshake.js

```diff
@@ -0,0 +1,409 @@
+ */
+function selectQuery(store, keyRange, resultMapper) {
+    const query = store.openCursor(keyRange);
+    return new Promise((resolve, reject) => {
+        let results = [];
+        query.onerror = (event) => {
+            reject(new Error("Query failed: " + event.target.errorCode));
+        };
+        // collect results
+        query.onsuccess = (event) => {
+            const cursor = event.target.result;
+            if (!cursor) {
+                resolve(results);
+                return; // end of results
+            }
+            results.push(resultMapper(cursor));
+            cursor.continue();
+        }
+    });
+}
+
+
+let store = null;
+let logger = null;
+let initPromise = null;
+let bugReportEndpoint = null;
+module.exports = {
+
+    /**
+     * Configure rage shaking support for sending bug reports.
+     * Modifies globals.
+     * @return {Promise} Resolves when set up.
+     */
+    init: function() {
+        if (initPromise) {
+            return initPromise;
+        }
+        logger = new ConsoleLogger();
+        if (window.indexedDB) {
+            store = new IndexedDBLogStore(window.indexedDB, logger);
+            initPromise = store.connect();
+            return initPromise;
+        }
+        initPromise = Promise.resolve();
+        return initPromise;
+    },
+
+    /**
+     * Force-flush the logs to storage.
+     * @return {Promise} Resolved when the logs have been flushed.
+     */
+    flush: async function() {
+        if (!store) {
+            return;
+        }
+        await store.flush();
+    },
+
+    /**
+     * Clean up old logs.
+     * @return Promise Resolves if cleaned logs.
+     */
+    cleanup: async function() {
+        if (!store) {
+            return;
+        }
+        await store.consume(false);
+    },
+
+    setBugReportEndpoint: function(url) {
+        bugReportEndpoint = url;
+    },
+
+    /**
+     * Send a bug report.
+     * @param {string} userText Any additional user input.
+     * @return {Promise} Resolved when the bug report is sent.
+     */
+    sendBugReport: async function(userText) {
+        if (!logger) {
+            throw new Error("No console logger, did you forget to call init()?");
+        }
+        if (!bugReportEndpoint) {
+            throw new Error("No bug report endpoint has been set.");
+        }
+
+        let version = "UNKNOWN";
+        try {
+            version = await PlatformPeg.get().getAppVersion();
+        }
+        catch (err) {} // PlatformPeg already logs this.
+
+        let userAgent = "UNKNOWN";
+        if (window.navigator && window.navigator.userAgent) {
+            userAgent = window.navigator.userAgent;
+        }
+
+        // If in incognito mode, store is null, but we still want bug report sending to work going off
+        // the in-memory console logs.
+        let logs = [];
+        if (store) {
+            logs = await store.consume(true);
```

Comment:
I don't really intend for `flush()` to ever be called outside this class, so it's dubious as to whether I should keep it in or not. You can never get duplicates because `flush()` wipes the `ConsoleLogger.logs` so only one function will win at getting the logs. It's possible that if `rageshake.flush()` is called and then `sendBugReport()` is called and `store.consume()` finishes before `rageshake.flush()` does, that you can get a missing chunk. Will guard.